### PR TITLE
Serverless docs rewrite/restructure part 2 (workers + handlers + vllm)

### PR DESCRIPTION
**Pull request by muhsinking**
## Overview

This PR implements a major restructuring of the Serverless documentation and architecture, along with several other miscellaneous updates and improvements:

1. Rewrote the `/serverless/workers/` and `/serverless/vllm/` folders
2. Consolidated `/serverless/handlers/` into `/serverless/workers/`
3. Deleted `/serverless/quick-deploy` and merged the instructions into `/serverless/overview`
4. Added a new deployment section to `/serverless/overview`: "Fork a worker template"
5. Moved `/runpodctl/` out of the main documentation TOC, into a new tab in the top nav
6. Rewrote `/get-started/api-keys`

### Workers Folder Rewrite
- Rewrote worker overview
- Created detailed guides for GitHub integration and worker deployment
- Updated worker configuration documentation with clearer terminology
- Improved organization of worker-related tutorials and examples
- Improved navigation structure with updated category positions

### Handlers Consolidation
- Moved all handler-related content into the `/serverless/workers/` directory
- Consolidated handler function documentation into a single comprehensive guide
- Added a comprehensive tutorial for concurrent handlers
- Updated all references to handlers throughout the documentation

### vLLM Folder Rewrite
- Rewrote all vLLM pages
- Deleted environment variables page; referred users to the GitHub README (don't want to maintain information in multiple places)
- Rewrote get-started page as an end-to-end tutorial using quick deploys
- Created new page for sending RunPod API requests to vLLM endpoints
- Expanded OpenAI API compatibility guide

## Impact
These changes significantly improve the organization and clarity of the Serverless documentation, making it easier for users to:

- Navigate through worker-related documentation
- Implement different types of handlers
- Deploy and manage vLLM workers

## Testing
All documentation links and references have been verified to ensure proper navigation and consistency with the new structure.

### CODE REVIEW

**Review by TimPietruskyRunPod — CHANGES_REQUESTED** [review]
this is already super awesome, thank you very much @muhsinking! 

## i love

- "handler functions" are combined with a separate page for the concurrent handlers, makes total sense
- the super clean toc <3 

## some ideas to further improve the docs

* on serverless/overview: i would move the "key concepts" before the "deployment" section, as it's important to understand on a high level what "endpoint" vs "worker" are (and maybe "handler functions" as a child of worker instead of using the same headline-level). and we might want to add a talk i have given about "what is serverless" https://www.youtube.com/watch?v=jzHIVyOhDlE, as it might be good for people to just watch something instead of only reading
* in the toc: i would put the "endpoint" first (because that's the order from "key concepts", then "worker"
* maybe it would make sense to have "serverless/deployment" and there we put everything that can be deployed? like "quick deploy an endpoint" or "vLLM workers"? and then maybe turn "vLLM workers" also into more like a tutorial style "thing"? "deploy an LLM with vLLM" (or what ever you think makes sense here, i just try to give ideas :D)

File: docs/serverless/vllm/get-started.md [hunk 1]
```diff
@@ -1,265 +1,161 @@
...
-3. (Optional) Add a Hugging Face token for gated models.
-4. Review your options and choose **Next**.
+- Llama 2 (e.g., `meta-llama/Llama-2-7b-chat-hf`)
+- Mistral (e.g., `mistralai/Mistral-7B-Instruct-v0.2`)
+- OpenChat (e.g., `openchat/openchat-3.5-0106`)
+- Gemma (e.g., `google/gemma-7b-it`)
+- And many more
 
-On the **vLLM parameters** page, provide additional parameters and options for your model:
+For this walkthrough, we'll use `openchat/openchat-3.5-0106`, but you can substitute this with any compatible model.
```
Comment (Yhlong00) [inline docs/serverless/vllm/get-started.md:38]: Maybe we should provide an updated list of popular small models, the current one feels a bit outdated.

File: docs/serverless/vllm/get-started.md [hunk 2]
```diff
@@ -1,265 +1,161 @@
...
-### Python Example
+If you need to customize your model deployment, you can edit your endpoint settings to add environment variables. Here are some useful environment variables you might want to set:
 
-```python
-import requests
+- `MAX_MODEL_LEN`: Maximum context length (e.g., `16384`)
+- `DTYPE`: Data type for model weights (`float16`, `bfloat16`, or `float32`)
+- `GPU_MEMORY_UTILIZATION`: Controls VRAM usage (e.g., `0.95` for 95%)
+- `CUSTOM_CHAT_TEMPLATE`: For models that need a custom chat template
+- `OPENAI_SERVED_MODEL_NAME_OVERRIDE`: Change the model name to use in OpenAI requests
```
Comment (Yhlong00) [inline docs/serverless/vllm/get-started.md:133]: https://github.com/runpod-workers/worker-vllm?tab=readme-ov-file#environment-variablessettings
a full list of env

File: docs/serverless/vllm/get-started.md [hunk 3]
```diff
@@ -1,265 +1,161 @@
...
-                 "prompt": "Write a haiku about nature.",
-                 "sampling_params": {"temperature": 0.8, "max_tokens": 50}
-         }}'
-```
+If you encounter issues with your deployment:
 
----
+- **Worker fails to initialize**: Check that your model is compatible with vLLM and your GPU has enough VRAM
+- **Slow response times**: Consider using a more powerful GPU or optimizing your request parameters
+- **Out of memory errors**: Reduce the max_tokens parameter or use a smaller model
```
Comment (Yhlong00) [inline docs/serverless/vllm/get-started.md:149]: max_token controls the maximum number of output tokens, while the MAX_CONCURRENCY environment variable defines how many requests can be processed in parallel.

You can refer to this link for details on how to use max_token:
https://github.com/runpod-workers/worker-vllm?tab=readme-ov-file#text-input-formats

By default, if max_token isn’t set, the output is limited to just 16 tokens,  this has been a common issue for many users, so it’s important to explicitly set it when sending your requests.

File: docs/serverless/vllm/overview.md [hunk 4]
```diff
@@ -1,60 +1,73 @@
...
 
-You can deploy most [models from Hugging Face](https://huggingface.co/models?other=LLM).
-For a full list of supported models architectures, see [Compatible model architectures](https://github.com/runpod-workers/worker-vllm/blob/main/README.md#compatible-model-architectures).
+There are two ways to deploy a vLLM worker:
 
-## Getting started
+**Quick deploy a vLLM endpoint**: The simplest approach—use RunPod's UI to deploy a model directly from Hugging Face with minimal configuration. For step-by-step instructions, see [Deploy a vLLM worker](/serverless/vllm/get-started).
 
-At a high level, you can set up the vLLM Worker by:
+**Deploy using a prebuilt worker image**: Deploy a preconfigured vLLM worker image from [GitHub](https://github.com/runpod-workers/worker-vllm) or [Docker Hub](https://hub.docker.com/r/runpod/worker-v1-vllm/tags), configuring your endpoint using [environment variables](https://github.com/runpod-workers/worker-vllm#environment-variablessettings).
```
Comment (Yhlong00) [inline docs/serverless/vllm/overview.md:29]: By default, if you’re using the VLLM quick deploy, the worker will need to download the model during initialization, which can take some time depending on the model selected.
If you’re running a production VLLM endpoint, it’s best to package the model into the Docker image ahead of time. This significantly reduces cold start time.

User can follow the instructions here to build the model into your image:
https://github.com/runpod-workers/worker-vllm?tab=readme-ov-file#option-2-build-docker-image-with-model-inside

File: docs/serverless/workers/concurrent-handler.md [hunk 5]
```diff
@@ -0,0 +1,169 @@
...
+    # Increase concurrency if under max limit and request rate is high
+    if (request_rate > high_request_rate_threshold and 
+        current_concurrency < max_concurrency):
+        return current_concurrency + 1
+    # Decrease concurrency if above min limit and request rate is low
+    elif (request_rate <= high_request_rate_threshold and 
+          current_concurrency > min_concurrency):
+        return current_concurrency - 1
+    
+    return current_concurrency
```
Comment (Yhlong00) [inline docs/serverless/workers/concurrent-handler.md:96]: Many users are having trouble understanding how this function works. It might help to start with a simple, minimal example, like:

```
def adjust_concurrency(current_concurrency):
    return 50
```

Then follow it with a more complex, real-world example like the one above. This way, users can build their understanding step by step.

File: docs/serverless/workers/deploy.md [hunk 6]
```diff
@@ -1,117 +1,168 @@
...
-Running any Action that sends requests to RunPod occurs a cost.
+# Copy your code and model files
+COPY src/handler.py .
+// highlight-start
+COPY models/ /models/
 
-:::
+# Set environment variables if needed
+ENV MODEL_PATH=/models/my_model.pt
+// highlight-end
```
Comment (Yhlong00) [inline docs/serverless/workers/deploy.md:84]: this one seems not looking good on dark mode:
image

File: docs/serverless/workers/handler-functions.md [hunk 7]
```diff
@@ -0,0 +1,289 @@
...
+
+When implementing async handlers, ensure proper use of `async` and `await` keywords throughout your code to maintain truly non-blocking operations and prevent performance bottlenecks, and consider leveraging the `yield` statement to generate outputs progressively over time.
+
+Always test your async code thoroughly to properly handle asynchronous exceptions and edge cases, as async error patterns can be more complex than in synchronous code.
+
+:::
+
+### Concurrent handlers
+
+Concurrent handlers process multiple requests simultaneously with a single worker. Use these for I/O-bound operations like API calls, database queries, or file operations where a single worker can efficiently manage multiple tasks at once.
```
Comment (Yhlong00) [inline docs/serverless/workers/handler-functions.md:157]: The concurrent handler is ideal for small, fast tasks that don’t fully utilize the entire GPU

**Review by Yhlong00 — DISMISSED** [review]

File: docs/runpodctl/overview.md
Comment (brendanmckeag) [inline docs/runpodctl/overview.md:1]: You can actually create and update serverless endpoints as well as pods the same way with the CLI - it's not limited to projects

File: docs/serverless/endpoints/endpoint-configurations.md
Comment (brendanmckeag) [inline docs/serverless/endpoints/endpoint-configurations.md:1]: My understanding of flashboot is that it keeps resources around after they're no longer used and if they're needed again before some other process claims them we can just yank them back to life - this might be a bit more better explained than "probabilistically" (which is actually still true, but might raise more questions than answers in the mind of the reader)

File: docs/serverless/vllm/get-started.md (hunk 1 above)
Comment (muhsinking) [reply docs/serverless/vllm/get-started.md:38]: @TimPietruskyRunPod @brendanmckeag What would you say are the most popular small models today?

File: docs/serverless/workers/deploy.md (hunk 6 above)
Comment (muhsinking) [reply docs/serverless/workers/deploy.md:84]: Oof that's annoying—I'll just remove the highlight for now, and maybe add it back after we've migrated to Mintlify ^-^;

File: docs/serverless/workers/concurrent-handler.md (hunk 5 above)
Comment (muhsinking) [reply docs/serverless/workers/concurrent-handler.md]: Added a step to build this and explain it in more detail!

File: docs/serverless/endpoints/endpoint-configurations.md
Comment (muhsinking) [reply docs/serverless/endpoints/endpoint-configurations.md:1]: Expanded this, lmk what you think of the new explanation!

File: docs/runpodctl/overview.md
Comment (muhsinking) [reply docs/runpodctl/overview.md:1]: Good to know! Made mention of this.
